src/wp-includes/class-wp-block.php

```diff
@@ -333,6 +333,60 @@ private function replace_html( string $block_content, string $attribute_name, $s
 		switch ( $block_type->attributes[ $attribute_name ]['source'] ) {
 			case 'html':
 			case 'rich-text':
+				if ( 'core/image' === $this->name && 'caption' === $attribute_name ) {
+					// Create private anonymous class until the HTML API provides `set_inner_html` method.
+					$bindings_processor = new class( $block_content, WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE ) extends WP_HTML_Processor {
+						/**
+						 * Replace the inner content of a figcaption element with the passed content.
+						 *
+						 * DO NOT COPY THIS METHOD.
+						 * THE HTML PROCESSOR WILL HAVE A PROPER METHOD.
+						 * USE IT INSTEAD.
+						 *
+						 * @since 6.7.0
+						 *
+						 * @param string $new_content New content to insert in the figcaption element.
+						 * @return bool Whether the inner content was properly replaced.
+						 */
+						public function set_figcaption_inner_text( $new_content ) {
+							// Check that the processor is paused on an opener tag.
+							if ( 'FIGCAPTION' !== $this->get_tag() || $this->is_tag_closer() ) {
+								return false;
+							}
+
+							// Once this element closes the depth will be one shallower than it is now.
+							$depth = $this->get_current_depth();
+							while ( $this->next_token() && $this->get_current_depth() >= $depth ) {
+								// This is inside the FIGCAPTION element.
+							}
+
+							if ( null !== $this->get_last_error() || $this->paused_at_incomplete_token() ) {
+								return false;
+							}
+
+							$this->set_bookmark( 'here' );
+
+							$opening = $this->bookmarks[ $this->current_element->token->bookmark_name ];
+							$closing = $this->bookmarks['_here'];
+							$start   = $opening->start + $opening->length;
+
+							$this->lexical_updates[] = new WP_HTML_Text_Replacement(
+								$start,
+								$closing->start - $start,
+								wp_kses_post( $new_content )
```

Comment:
if inner text is all we care about we can use `esc_html()` instead of `wp_kses_post()` and that would be a bit more reliable, but if we tried to do something like add a link inside the caption then it will be escaped as `<a href=…>` instead of rendering as a link.

Reply:
I just checked and you can actually add a link in a caption, so I assume we should keep using `wp_kses_post()`. I have renamed the function to say html instead of text.